### PR TITLE
Update the architecture about http requests and logins/outs

### PR DESCRIPTION
**Pull request by VitSalis**
* Now the real-time service can accept http requests from the client
* From now on the real-time service will forward messages to the
  RESTful API.
* The real-time API will now check if a user is valid by checking
  his ting_auth token, which can be retrieved from the RESTful API.

### CODE REVIEW

**Comment by dionyziz** [comment]
Please clean up your commits by splitting them into atomic ones.

File: etc/spec/ARCHITECTURE.md [hunk 1]
```diff
@@ -47,38 +47,32 @@ privileged: The privileges of the real-time service are elevated so that
...
-     the recipient of the message is a channel or a user.
-  2. `target`: The name of the channel or the user we wish to send the message
-     to.
-  3. `text`: The text of the message.
+## Real-time API
 
-* `start-typing`: Indicates that a user started typing a new message. Expects
-  the following parameters:
+The real-time API deals with maintaining the state of who is online and performs
+message exchange. Also it acts as a proxy for the RESTful API. The real-time
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: It also acts as a proxy...

File: etc/spec/ARCHITECTURE.md [hunk 2]
```diff
@@ -47,38 +47,32 @@ privileged: The privileges of the real-time service are elevated so that
...
+API decides which operations will be handled by it and which will be forwarded to
+the Django RESTful API.
+Communication happens using both HTTP and web sockets. Web socket
+communication is implemented using [socket.io](http://socket.io/).
 
-  1. `type`: A string which is either `channel` or `user`.
-  2. `target`: The name of the channel or the user we wish to send the message
-     to.
-  3. `text`: The text of the message typed so far.
+The real-time server allows the following web-socket messages to be sent:
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: to be sent to it

File: etc/spec/ARCHITECTURE.md [hunk 3]
```diff
@@ -47,38 +47,32 @@ privileged: The privileges of the real-time service are elevated so that
...
+* `login`: Indicates that a user is logging in on the server by providing the
+  user information. Requires a `username` and a `ting_auth` parameter. The real-time
+  service then checks that the authentication tokens are valid with the RESTful
+  API. If the authentication fails, the real-time service closes the connection.
+  The `ting_auth` parameter provided by the client must be gathered by the client
+  by requesting a new session from the RESTful API. This message must be sent before
+  any other messages.
 
-The server can publish the following messages:
+The real-time server can publish using web-sockets the following messages:
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: the following messages using web sockets

**Comment by dionyziz** [comment]
As of this change, we do not need to have privileged operations. We can simply put django behind a firewall and only expose the URLs that are public. Hence, the paragraphs about password protection can go away and be replaced with a mention that the django service is not public.

Then, for each of the django operations, you must mention whether or not it will be reverse proxied by node.

File: etc/spec/ARCHITECTURE.md [hunk 4]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
   4. `datetime_start`: A unix epoch in ms which indicates the datetime the
      message started being typed.
 
+The real-time server exposes an HTTP endpoint at `https://ting.gr/api/v1` with the following
+URLs:
+
+* A PATCH operation on `/message`: Sends an update on the message with an id equal to <messageid>.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: I don't see <messageid> on the URL.

File: etc/spec/ARCHITECTURE.md [hunk 5]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+  The message can be sent from the user to a channel or to another user directly.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: Please assume the reader of ARCH has read the SPEC. This sentence here is needless and confusing – are you talking about a ting message, a piece of text exchanged between users and channels? Then this should be on the SPEC. Are you talking about a protocol message, as should be on the ARCH? Then how can a channel receive such messages?

I suggest you remove this sentence completely.

File: etc/spec/ARCHITECTURE.md [hunk 6]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+  Takes five parameters:
+
+  1. `type`: A string which is either `channel` or `user`, indicating whether
+     the recipient of the message is a channel or a user.
+  2. `target`: The name of the channel or the user we wish to send the message
+     to.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: Why do you need these for patching a message whose id is already known?

File: etc/spec/ARCHITECTURE.md [hunk 7]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+  3. `text`: The text of the message.
+  4. `typing`: A boolean that indicates if the message is currently
+     being typed or not.
+  5. `messageid`: An integer which is the unique id of the message being sent.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: Why don't you pass this in the URL for uniformity? This is how we address messages anyway.

File: etc/spec/ARCHITECTURE.md [hunk 8]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+
+     * There is no response for this operation.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: You should say that the response is a No Content response.

Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: Also please clarify what happens when there is some validation problem with the data. What sort of response should be expected?

File: etc/spec/ARCHITECTURE.md [hunk 9]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+
+* A POST operation on `/message`: Indicates that a user started typing a new message. Expects
+  the following parameters:
+
+  1. `type`: A string which is either `channel` or `user`.
+  2. `target`: The name of the channel or the user we wish to send the message
+     to.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: These should be in the URL

File: etc/spec/ARCHITECTURE.md [hunk 10]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+  3. `text`: The text of the message typed so far.
+
+    * The response contains the `messageid`, which is the id of the
+      new message that is currently being typed.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: How does it contain it? Is it JSON? Is it just the id? I think it should be JSON for uniformity with the rest of the API.

File: etc/spec/ARCHITECTURE.md [hunk 11]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+
+     * There is no response for this operation.
+
+* A POST operation on `/message`: Indicates that a user started typing a new message. Expects
+  the following parameters:
+
+  1. `type`: A string which is either `channel` or `user`.
+  2. `target`: The name of the channel or the user we wish to send the message
+     to.
+  3. `text`: The text of the message typed so far.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: You need `typing` here.

File: etc/spec/ARCHITECTURE.md [hunk 12]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+
+    * The response contains the `messageid`, which is the id of the
+      new message that is currently being typed.
+
+* Operations on `/messages/` and `/channels/` are forwarded to the RESTful API
+  with the appropriate parameters..
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: nit double dot

File: etc/spec/ARCHITECTURE.md [hunk 13]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+  2. `target`: The name of the channel or the user we wish to send the message
+     to.
+  3. `text`: The text of the message.
+  4. `typing`: A boolean that indicates if the message is currently
+     being typed or not.
+  5. `messageid`: An integer which is the unique id of the message being sent.
+
+     * There is no response for this operation.
+
+* A POST operation on `/message`: Indicates that a user started typing a new message. Expects
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: This should be on `/meesages`, not `/message`, for consistency with the rest of the API. e.g. `/users` and `/sessions`. This is also the correct way semantically: A POST operation is creating a resource under a path; and that path is `/messages`, not `/message`. This is similar to a file creation operation within a specified directory: You should be creating the file under the same directory which you will be reading it from.

File: etc/spec/ARCHITECTURE.md [hunk 14]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
...
+* Operations on `/messages/` and `/channels/` are forwarded to the RESTful API
+  with the appropriate parameters..
+
 ## RESTful API
 
 The RESTful API deals with four resources: Messages, Channels, Users, Sessions. The
 responses are always given in JSON. As such, we make no use of Django templates,
 only models and views. The URLs of the RESTful API live under the
-`https://ting.gr/api/v1` URL.
+`https://ting.gr/restapi/v1` URL.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: This will no longer be a public URL

File: etc/spec/ARCHITECTURE.md [hunk 15]
```diff
@@ -254,7 +277,20 @@ request, it gets created internally.
...
 with the session, the user gets deleted, too.
+
+####Checking authentication
+
+This operation is performed by the real-time service to the RESTful API server for verifying
+that a given client cookie is correct.
+
+A GET request on `/sessions/<username>`. This request must be performed using the `ting_auth`
+cookie that the user provided. If the cookie is valid for the given user, a 204 (No Content) response
+is returned. Otherwise, a 403 (Forbidden) response is returned.
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: Please follow the styling conventions from the `sessions` section on status codes for consistency. I didn't look at those portions when writing this, but we should aim to remain uniform. "the server returns a 204 status code" seems to be the language we're using there.

File: etc/spec/ARCHITECTURE.md [hunk 16]
```diff
@@ -119,12 +109,44 @@ The server can publish the following messages:
   4. `datetime_start`: A unix epoch in ms which indicates the datetime the
      message started being typed.
 
+The real-time server exposes an HTTP endpoint at `https://ting.gr/api/v1` with the following
+URLs:
+
+* A PATCH operation on `/message`: Sends an update on the message with an id equal to <messageid>.
+  The message can be sent from the user to a channel or to another user directly.
+  Takes five parameters:
```
Comment (dionyziz) [inline etc/spec/ARCHITECTURE.md]: Be consistent. "Expects the following parameters" VS "Takes five parameters".

**Comment by dionyziz** [comment]
As there is now no communication between the client and the RESTful API service, there needs to be a change in the diagram. The arrow between the client and the Django service must be removed.

Also, since there is HTTPS communication between the client and the node service, a new arrow must be added to the diagram for that.

File: etc/spec/ARCHITECTURE.md (hunk 6 above)
Comment (VitSalis) [inline etc/spec/ARCHITECTURE.md]: Yes, you're right.

**Comment by dionyziz** [comment]
This is stalling do I'm closing this. If you intend to work on it please reopen.

**Pull request closed without merging**